src/libraries/Common/src/System/Security/Cryptography/ECDsaOpenSsl.cs

```diff
@@ -105,17 +121,21 @@ public override bool TrySignHash(ReadOnlySpan<byte> hash, Span<byte> destination
                 out bytesWritten);
         }
 
+        // SafeEvpPKeyCtxHandle doesn't touch ref counts of SafeEvpPKeyHandle so we need to keep it alive during it's lifetime.
+        // We only use CreateCtx from within single `using` statement so we can guarantee no lifetime issues.
```

Comment:
Oh, this is bad.

A parallel call with Sign to Dispose or Import\* will now have us doing a UAF in native code.

This absolutely has to be written to either up-ref the key or (preferred) to never send the EVP_PKEY_CTX* back up to managed code.

Reply:
Won't the same problem occur if we *just* pass in the key to native and not use EVP_PKEY_CTX? We'd still need to upref the key but we'd do that in native before passing in to native but we won't be able to check if key is disposed? I didn't realize we guarantee thread safety on these types but since this is almost for free (upref) I will change it.
I really do like the EVP_PKEY_CTX being done in managed as it's much nicer to debug than native and I don't see any strong reason to have more native code (the single EvpPKeyHandle allocation doesn't seem to be bad comparing to everything happening here and it also seem to make it less error prone).

I'm not sure I understand why would we want to have more native code

Reply:
I have addressed the up-ref/UAF comment but didn't get rid of EVP_PKEY_CTX*

Reply:
> Won't the same problem occur if we just pass in the key to native?

No, because the P/Invoke boundary will have done DangerousAddRef, and keeps the key alive until the function returns.